File: go/ontology/similarity.go
// SPDX-License-Identifier: Apache-2.0
package ontology

import (
	"fmt"
	"math"
	"strings"
)

// JaroWinklerDistance computes case-insensitive Jaro-Winkler similarity
// between two strings. Returns a value in [0, 1] where 1 indicates
// identical strings and 0 indicates completely dissimilar strings.
//
// This is a native implementation with no external dependencies.
//
// Time: O(max(len(s1), len(s2)))
// Space: O(max(len(s1), len(s2)))
func JaroWinklerDistance(s1, s2 string) float64 {
	a := strings.TrimSpace(strings.ToLower(s1))
	b := strings.TrimSpace(strings.ToLower(s2))
	if a == "" && b == "" {
		return 1.0
	}
	if a == "" || b == "" {
		return 0.0
	}
	if a == b {
		return 1.0
	}

	jaro := jaroDistance(a, b)

	ra := []rune(a)
	rb := []rune(b)
	commonPrefix := 0
	limit := 4
	if len(ra) < limit {
		limit = len(ra)
	}
	if len(rb) < limit {
		limit = len(rb)
	}
	for i := 0; i < limit; i++ {
		if ra[i] != rb[i] {
			break
		}
		commonPrefix++
	}

	return jaro + float64(commonPrefix)*0.1*(1.0-jaro)
}

// jaroDistance computes the Jaro similarity between two strings.
// Both strings must be non-empty and already lowercased/trimmed.
// Uses []rune internally for correct multi-byte character handling.
func jaroDistance(s1, s2 string) float64 {
	r1 := []rune(s1)
	r2 := []rune(s2)

	maxLen := len(r1)
	if len(r2) > maxLen {
		maxLen = len(r2)
	}
	matchWindow := maxLen/2 - 1
	if matchWindow < 0 {
		matchWindow = 0
	}

	r1Matches := make([]bool, len(r1))
	r2Matches := make([]bool, len(r2))

	matches := 0
	transpositions := 0

	for i := 0; i < len(r1); i++ {
		start := i - matchWindow
		if start < 0 {
			start = 0
		}
		end := i + matchWindow + 1
		if end > len(r2) {
			end = len(r2)
		}
		for j := start; j < end; j++ {
			if r2Matches[j] || r1[i] != r2[j] {
				continue
			}
			r1Matches[i] = true
			r2Matches[j] = true
			matches++
			break
		}
	}

	if matches == 0 {
		return 0.0
	}

	k := 0
	for i := 0; i < len(r1); i++ {
		if !r1Matches[i] {
			continue
		}
		for !r2Matches[k] {
			k++
		}
		if r1[i] != r2[k] {
			transpositions++
		}
		k++
	}

	return (float64(matches)/float64(len(r1)) +
		float64(matches)/float64(len(r2)) +
		float64(matches-transpositions/2)/float64(matches)) / 3.0
}

// CosineSimilarity computes the cosine similarity between two float32
// vectors. Returns a value in [-1, 1] where 1 indicates identical
// direction, 0 indicates orthogonal vectors, and -1 indicates opposite
// direction. Returns 0 when either vector has zero magnitude.
//
// Returns an error if the vectors have different lengths.
//
// Time: O(n) where n = len(a)
// Space: O(1)
func CosineSimilarity(a, b []float32) (float64, error) {
	if len(a) != len(b) {
		return 0, fmt.Errorf("ontology: cosine similarity requires equal-length vectors, got %d and %d", len(a), len(b))
	}
	if len(a) == 0 {
		return 0, nil
	}

	var dot, normA, normB float64
	for i := range a {
		ai := float64(a[i])
		bi := float64(b[i])
		dot += ai * bi
		normA += ai * ai
		normB += bi * bi
	}

	magA := math.Sqrt(normA)
	magB := math.Sqrt(normB)
	if magA == 0 || magB == 0 {
		return 0, nil
	}
	return dot / (magA * magB), nil
}

File: sdks/ts/memory/src/ontology/similarity.ts
// SPDX-License-Identifier: Apache-2.0

/**
 * String and vector similarity functions for the ontology deduplication
 * system. Jaro-Winkler is implemented inline (~30 lines) to avoid adding
 * a dependency for a single algorithm.
 */

/**
 * Compute the Jaro similarity between two strings.
 *
 * Time: O(max(len(s1), len(s2)))
 * Space: O(max(len(s1), len(s2)))
 */
function jaroSimilarity(s1: string, s2: string): number {
  if (s1.length === 0 && s2.length === 0) return 1.0
  if (s1.length === 0 || s2.length === 0) return 0.0

  const matchWindow = Math.max(0, Math.floor(Math.max(s1.length, s2.length) / 2) - 1)
  const s1Matches = new Array<boolean>(s1.length).fill(false)
  const s2Matches = new Array<boolean>(s2.length).fill(false)

  let matches = 0
  let transpositions = 0

  for (let i = 0; i < s1.length; i++) {
    const start = Math.max(0, i - matchWindow)
    const end = Math.min(i + matchWindow + 1, s2.length)
    for (let j = start; j < end; j++) {
      if (s2Matches[j] || s1[i] !== s2[j]) continue
      s1Matches[i] = true
      s2Matches[j] = true
      matches++
      break
    }
  }

  if (matches === 0) return 0.0

  let k = 0
  for (let i = 0; i < s1.length; i++) {
    if (!s1Matches[i]) continue
    while (!s2Matches[k]) k++
    if (s1[i] !== s2[k]) transpositions++
    k++
  }

  return (
    (matches / s1.length + matches / s2.length + (matches - transpositions / 2) / matches) / 3
  )
}

/**
 * Compute the Jaro-Winkler similarity between two strings.
 * Case-insensitive and whitespace-trimmed. Returns a value in [0, 1]
 * where 1 indicates identical strings.
 *
 * Time: O(max(len(s1), len(s2)))
 * Space: O(max(len(s1), len(s2)))
 */
export function jaroWinklerDistance(s1: string, s2: string): number {
  const a = s1.trim().toLowerCase()
  const b = s2.trim().toLowerCase()
  if (a.length === 0 && b.length === 0) return 1.0
  if (a.length === 0 || b.length === 0) return 0.0

  const jaro = jaroSimilarity(a, b)

  let commonPrefix = 0
  const maxPrefix = Math.min(4, Math.min(a.length, b.length))
  for (let i = 0; i < maxPrefix; i++) {
    if (a[i] !== b[i]) break
    commonPrefix++
  }

  return jaro + commonPrefix * 0.1 * (1 - jaro)
}

/**
 * Compute the cosine similarity between two number arrays. Returns a
 * value in [-1, 1] where 1 indicates identical direction. Returns 0
 * when either vector has zero magnitude.
 *
 * Throws if the vectors have different lengths.
 *
 * Time: O(n) where n = a.length
 * Space: O(1)
 */
export function cosineSimilarity(a: readonly number[], b: readonly number[]): number {
  if (a.length !== b.length) {
    throw new Error(`ontology: cosine similarity requires equal-length vectors (${a.length} vs ${b.length})`)
  }
  if (a.length === 0) return 0

  let dot = 0
  let normA = 0
  let normB = 0

  for (let i = 0; i < a.length; i++) {
    const ai = a[i]!
    const bi = b[i]!
    dot += ai * bi
    normA += ai * ai
    normB += bi * bi
  }

  const magA = Math.sqrt(normA)
  const magB = Math.sqrt(normB)
  if (magA === 0 || magB === 0) return 0

  return dot / (magA * magB)
}
